README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

[![Project Status: Active – The project has reached a stable, usable
state and is being actively
developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![R build
status](https://github.com/cont-limno/LAGOSUS/workflows/R-CMD-check/badge.svg)](https://github.com/cont-limno/LAGOSUS/actions)
[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/LAGOSUS)](https://cran.r-project.org/package=LAGOSUS)
[![CRAN RStudio mirror
downloads](http://cranlogs.r-pkg.org/badges/LAGOSUS)](https://cran.r-project.org/package=LAGOSUS)

# LAGOSUS <img src="man/figures/logo.png" align="right" height=140/>

The `LAGOSUS` package provides an R interface to download LAGOS-US data,
store this data locally, and perform a variety of filtering and
subsetting operations.

LAGOS-US contains data for 479,950 lakes and reservoirs larger than 1 ha
in continental US. The database includes 4 data modules for: lake
location and physical characteristics for all lakes; ecological context
(i.e., the land use, geologic, climatic, and hydrologic setting of
lakes) for all lakes; in situ measurements of lake water quality for a
subset of the lakes from the past 3 decades for approximately Y-Z lakes
depending on the variable (see
[References](https://github.com/cont-limno/LAGOSUS#references)); and
depth for a subset of all lakes (forthcoming).

## Installation

``` r
# install development version from Github
# install devtools if not found - install.packages("devtools")
devtools::install_github("cont-limno/LAGOSUS", dependencies = TRUE)
```

### Data

General users currently have public access to the locus module. It can
be downloaded and stored to your local system with the following
command:

``` r
library(LAGOSUS)

# only the locus module is currently public
lagosus_get(dest_folder = lagosus_path())
```

Currently only the “locus” module of LAGOS-US has been released in a
public repository. Members of the development team who have access to
unreleased modules (limno, geo, etc), will need to use the the
`lagosus_compile` function (not `lagosus_get`) and supply the path to
their local `locus`, `limno`, `geo`, or `depth` data folders. Replace
the paths in the example below with the path to each respective folder
on your system. Most people will have access to these folders through
Dropbox. For example, the `locus_folder` would be assigned to something
like: `C:/Users/FWL/Dropbox/CL_LAGOSUS_exports/LAGOSUS_LOCUS`

<!-- dir("../../../Downloads/") -->

``` r
# an example for members of the dev team to specify local data folder paths
lagosus_compile(
  locus_version = "1.0",
  locus_folder = "~/Downloads/LAGOSUS_LOCUS/LOCUS_v1.0",
  limno_version = "2.1",
  limno_folder = "~/Downloads/LAGOSUS_LIMNO/US/LIMNO_v2.1/Final exports",
  depth_version = "0.1",
  depth_folder = "~/Downloads/LAGOSUS_DEPTH/DEPTH_v0.1",
  geo_version = "1.0",
  geo_folder = "~/Downloads/LAGOSUS_GEO/GEO_EXPORT_BETA_v1",
  dest_folder = lagosus_path())
```

## Usage

### Load Package

``` r
library(LAGOSUS)
```

### Load data

The `lagosus_load` function returns a named list of `data.frame`
objects. Use the `names()` function to see a list of available data
frames `names(lg)`.

``` r
lg <- lagosus_load(modules = c("locus"))
names(lg)
```

<!-- ```{r load_data_cached, eval=FALSE, echo=FALSE} -->
<!-- dt <- readRDS(system.file("lagos_test_subset.rds", package = "LAGOSUS")) -->
<!-- names(dt) -->
<!-- ``` -->
<!-- #### Locate tables containing a variable  -->
<!-- ```{r eval=FALSE} -->
<!-- query_lagos_names("secchi") -->
<!-- ``` -->
<!-- ```{r echo=FALSE, eval=FALSE} -->
<!-- query_lagos_names("secchi", dt = dt) -->
<!-- ``` -->

#### Preview a table

``` r
head(lg$locus$lake_characteristics)
```

#### Preview a specific lake

``` r
lake_info(name = "Pine Lake", state = "California")
# or using a lagoslakeid
# lake_info(lagoslakeid = 4389)
```

#### Map specific lakes

``` r
library(mapview)
mapview(coordinatize(lake_info(name = "Pine Lake", state = "California")))
```

#### Read table metadata

``` r
# lookup which table(s) contain a column name
query_lagos_names("ws_meanwidth", dt = lg)
# load help file for a table
?lake_watersheds
```

<!-- ```{r load printr, echo=FALSE,message=FALSE,results='hide', eval=FALSE} -->
<!-- loadNamespace("printr") -->
<!-- ``` -->
<!-- ```{r Read metadata for individual tables, eval=FALSE} -->
<!-- help.search("datasets", package = "LAGOSUS") -->
<!-- ``` -->
<!-- ```{r unload printr, echo=FALSE, eval=FALSE} -->
<!-- unloadNamespace("printr") -->
<!-- ``` -->

## Legacy Versions

To install versions of `LAGOSUS` compatible with older versions of
LAGOS-US data modules, review the
[Changelog](https://cont-limno.github.io/LAGOSUS/news/index.html) to
find to package version associated with your desired module version. Run
the following command where `ref` is set to your desired version (in the
example, it is version 0.0.1):

``` r
# install devtools if not found
# install.packages("devtools")
devtools::install_github("cont-limno/LAGOSUS", ref = "v0.0.1")
```

## References

Cheruvelil, K.S., Soranno, P.A., McCullough, I.M., Webster, K.E.,
Rodriguez, L.K. and Smith, N.J., 2021. LAGOS‐US LOCUS v1. 0: Data module
of location, identifiers, and physical characteristics of lakes and
their watersheds in the conterminous US. Limnology and Oceanography
Letters, 6(5), pp.270-292.


NEWS.md
# LAGOSUS 0.0.4

* Added EDI data downloading functionality for the LOCUS module

## MODULE UPDATES

* Added support for the GEO module

# LAGOSUS 0.0.3

## MODULE UPDATES

 * Added support for the LIMNO module
 
 * Rolled back mistaken default version of LOCUS from 1.1 to 1.0

# LAGOSUS 0.0.2

## MODULE UPDATES

 * Updated DEPTH module from 0 to version 0.1

## BREAKING CHANGES

 * The names of nearly all data tables have changed to exactly match LAGOSUS schema (#11)

## MINOR IMPROVEMENTS

 * The `coordinatize` function now has a default projection (#12)
 
 * Added ability to specify module versions using an environment variable (#9)

# LAGOSUS 0.0.1

* Initial prototype release


R/LAGOSUS-package.R
#' Interface to the LAGOSUS lakes database
#' @name LAGOSUS-package
#' @aliases LAGOSUS
#' @docType package
#' @importFrom magrittr %>%
#' @title R interface to the LAGOSUS lakes database
#' @author \email{stachel2@msu.edu}
NULL

#' Latest LAGOSUS module versions
#'
#' @name lagosus_version
#' @details Defaults to the latest release version of each data module. Can be
#' overridden using the `LAGOSUS_LOCUS_VER`, `LAGOSUS_LIMNO_VER`,
#' `LAGOSUS_GEO_VER`, or `LAGOSUS_DEPTH_VER` environment variables; see example
#' below. Using different environment variables for the same module in a single
#' interactive session will not invalidate the `lagosus_load` cache. Manually
#' specify module version numbers if needing to load multiple different module
#' versions.
#' @export
#'
#' @examples \dontrun{
#' # use defaults
#' lagosus_version()
#'
#' # custom setting
#' Sys.setenv(LAGOSUS_DEPTH_VER = 0)
#' lagosus_version()
#'
#' # revert to defaults
#' Sys.setenv(LAGOSUS_DEPTH_VER = "")
#' lagosus_version()
#' }
lagosus_version <- function(){
  res_defaults <- data.frame(
    modules = c("locus", "limno", "geo", "depth"),
    versions = c("1.0","2.1","1.0","0.1"),
    stringsAsFactors = FALSE)

  env_vars <- Sys.getenv(
    paste0("LAGOSUS_", c("LOCUS", "LIMNO", "GEO", "DEPTH"), "_VER"), "")
  env_vars <- data.frame(
    modules = tolower(stringr::str_extract(names(env_vars),
                                           "(?<=LAGOSUS_)(.*)(?=_VER)")),
    versions = as.character(env_vars),
    stringsAsFactors = FALSE)

  if (any(!env_vars$versions == "")) {
    env_vars$defaults <- res_defaults$versions
    res               <- dplyr::mutate(env_vars,
                                       versions = dplyr::case_when(
                                         versions == "" ~ defaults,
                                         TRUE ~ versions))
  }else{
    res <- res_defaults
  }

  res
}

#' LAGOSUS Lake information
#'
#' Includes information about lake identifiers used in LAGOS-US, information
#' about lake location (i.e., lat, lon, and elevation), and the zone IDs for
#' each spatial division containing the lake (e.g., state, county, hydrologic
#' unit).
#'
#' @eval c("@details", get_table_metadata("locus", "lake_information"))
#'
#' @name lake_information
#' @docType data
#' @keywords datasets
NULL

#' LAGOSUS Lake characteristics
#'
#' Includes derived metrics for each lake such as lake geometry (e.g., area,
#'  perimeter, shape metrics), lake connectivity (e.g., lake connectivity
#'  class, number of upstream lakes), and whether the lake is located within an
#'  area that was glaciated during the Late Wisconsin glaciation.
#'
#' @eval c("@details", get_table_metadata("locus", "lake_characteristics"))
#'
#' @name lake_characteristics
#' @docType data
#' @keywords datasets
NULL

#' LAGOSUS Lake watersheds
#'
#' Includes identifiers, location, and geometry for the calculated watersheds (WS) and network
#' watersheds (NWS).
#'
#' @eval c("@details", get_table_metadata("locus", "lake_watersheds"))
#'
#' @name lake_watersheds
#' @docType data
#' @keywords datasets
#' @aliases watersheds
NULL

#' LAGOSUS Identifier links
#'
#' @description Includes single or multiple identifiers per lake from other commonly used
#'  national-scale data products including: Water Quality Portal (WQP),
#'  Geographic Names Information System (GNIS), NHD MR Plusv2, LAGOS-NE, and
#'  the EPA National Lake Assessment surveys from 2007 and 2012.
#'
#' @eval c("@details", get_table_metadata("locus", "lake_link"))
#'
#' @name lake_link
#' @docType data
#' @keywords datasets
NULL

#' LAGOSUS Data sources
#'
#' @eval get_table_metadata("locus", "source_table_locus")
#'
#' @name source_table_locus
#' @docType data
#' @keywords datasets
NULL

#' LAGOSUS Chemical and physical data
#'
#' @eval get_table_metadata("limno", "site_chemicalphysical")
#'
#' @name site_chemicalphysical
#' @docType data
#' @keywords datasets
NULL

#' LAGOSUS Contaminant data
#'
#' @eval get_table_metadata("limno", "site_contaminants")
#'
#' @name site_contaminants
#' @docType data
#' @keywords datasets
NULL

#' LAGOSUS LIMNO information
#'
#' @eval get_table_metadata("limno", "site_information")
#'
#' @name site_information
#' @docType data
#' @keywords datasets
NULL

#' LAGOSUS LIMNO Data sources
#'
#' @eval get_table_metadata("limno", "source_table_limno")
#'
#' @name source_table_limno
#' @docType data
#' @keywords datasets
NULL

#' LAGOSUS Depth data
#'
#' @eval c("@details", get_table_metadata("depth", "depth"))
#'
#' @name depth
#' @docType data
#' @keywords datasets
NULL

# #' LAGOSUS Terrain data
# #'
# #' @eval get_table_metadata("geo", "zone_terrain")
# #'
# #' @name zone_terrain
# #' @docType data
# #' @keywords datasets
# NULL

# #' LAGOSUS Soils data
# #'
# #' @eval get_table_metadata("geo", "zone_soils")
# #'
# #' @name zone_soils
# #' @docType data
# #' @keywords datasets
# NULL
